### PR TITLE
Fix more exploits (carpet dupe, rail dupe, bedrock destroying, falling block (sand/dragon egg/etc) duping, dragon exp dupe) and properly prevent tnt duplication

### PR DESCRIPTION
**Pull request by Spottedleaf**
Properly prevents physics desyncs by setting the old pos to air and by getting the old pos after the set type call

### CODE REVIEW

**Comment by Spottedleaf** [comment]
Marked as draft - needs further testing to confirm this does not break mechanics

**Comment by Spottedleaf** [comment]
This pr will be updated to fix a variety of other piston exploits (bedrock destroying)

**Comment by Spottedleaf** [comment]
Continuation of https://github.com/PaperMC/Paper/pull/3544

**Comment by Spottedleaf** [comment]
Updated to include a fix for portal sand duping

File: Spigot-Server-Patches/0543-fixup-Protect-Bedrock-and-End-Portal-Frames-from-bei.patch [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+     public co.aikar.timings.Timing getTiming() {
+diff --git a/src/main/java/net/minecraft/server/BlockPiston.java b/src/main/java/net/minecraft/server/BlockPiston.java
+index 66dbe85fe975defa783fb0eb5f6053eb2ad5c187..0c519c1cdd2deb4163ec6bb4a5aaf41d519b70f2 100644
+--- a/src/main/java/net/minecraft/server/BlockPiston.java
++++ b/src/main/java/net/minecraft/server/BlockPiston.java
+@@ -175,6 +175,12 @@ public class BlockPiston extends BlockDirectional {
+     @Override
+     public boolean a(IBlockData iblockdata, World world, BlockPosition blockposition, int i, int j) {
+         EnumDirection enumdirection = (EnumDirection) iblockdata.get(BlockPiston.FACING);
++        // Paper start - prevent retracting when we're facing the wrong way (we were replaced before retraction could occurr)
```
Comment (mbax) [inline Spigot-Server-Patches/0543-fixup-Protect-Bedrock-and-End-Portal-Frames-from-bei.patch]: "occurr"

File: Spigot-Server-Patches/0545-Prevent-position-desync-in-playerconnection-causing-.patch [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+index 7123e197c7ed01afd4fbf7aa0760611373039a13..69cec4647c27bca92b037f63ca513935f8bc037c 100644
+--- a/src/main/java/net/minecraft/server/PlayerConnection.java
++++ b/src/main/java/net/minecraft/server/PlayerConnection.java
+@@ -1080,6 +1080,11 @@ public class PlayerConnection implements PacketListenerPlayIn {
+                             }
+ 
+                             this.player.move(EnumMoveType.PLAYER, new Vec3D(d7, d8, d9));
++                            // Paper start - prevent position desync
++                            if (this.teleportPos != null) {
++                                return; // ... thanks mojang for letting move calls teleport accross dimensions.
```
Comment (mbax) [inline Spigot-Server-Patches/0545-Prevent-position-desync-in-playerconnection-causing-.patch]: "accross"

**Comment by Spottedleaf** [comment]
If the only issues we got are spelling we are ready for prod!

**Comment by Spottedleaf** [comment]
Updated to include dragon exp duping

**Comment by Spottedleaf** [comment]
Note: need to backport                 world.setTypeAndData(oldPos, Blocks.AIR.getBlockData(), 2 | 16 | 1024); // Paper - set air to prevent later physics updates from seeing this block

**Comment by TheMatt2** [comment]
Can you please make some of these protections toggleable in the config? I know on my server I consider lit TNT duping a feature, not a bug as it allows for mechanisms that can't be  accomplished any other way.

Additionally, I know a number of servers actually consider bedrock breaking a feature, as it allows for easy access to the nether roof.

To be clear, I think that these protections will be an excellent addition, but I think they are worthy of being toggleable.

**Comment by Spottedleaf** [comment]
Paper's policy when it comes to duping is to fix them without a config. See my previous PR.

However, for _bedrock breaking_ paper has a config - `allow-perm-block-break-exploits`. So in this PR I've also made the additional protections against bedrock breaking configurable under that option.

**Comment by TwoPointInfinity** [comment]
Where can I find paper's policy stating configs for patches aren't allowed?

**Comment by Spottedleaf** [comment]
Ready for review 😄 

The tnt duplication option has been renamed to the piston duplication option. The new piston duplication option defaults to the tnt duplication option if it exists. The new option controls tnt duplication, rail duplication, and carpet duplication.

**Comment by GNUGradyn** [comment]
I know It's been a while but paperMC really needs to reconsider their stance on making things configurable. Many advanced players want to push the game to its limits using unintended mechanics and therefore want the bugs as-is. It's fine if you don't agree, but there is literally no reason to refuse to accept to merge toggles for these unintended mechanics. Actively preventing people from playing the game how they want you because you think its "incorrect" or "not what mojang intended" is insane. Having the *option* to disable these checks wont affect the people who want to leave them on. It's not like you're saving development time - there have been countless PRs trying to fix this for you. It's not making the experience better for anyone, it's not saving development time, its just insane.

**Comment by Owen1212055** [comment]
Please read https://github.com/PaperMC/Paper/issues/3854

**Pull request merged**